diff --git a/include/board_hw_info.h b/include/board_hw_info.h
index 092a918..ee26ad2 100644
--- a/include/board_hw_info.h
+++ b/include/board_hw_info.h
@@ -49,6 +49,8 @@
 #define	BOARD_ID_SAMA5D4_XULT	17
 #define	BOARD_ID_SAMA5D36_CP	18
 #define	BOARD_ID_SAMA5D2_XULT	19
+#define BOARD_ID_SAMA5D4_LORIX_ONE_512	98
+#define BOARD_ID_SAMA5D4_LORIX_ONE	99
 
 /* Vendor id definition */
 #define VENDOR_EMBEST		1
@@ -57,6 +59,7 @@
 #define VENDOR_COGENT		4
 #define VENDOR_PDA		5
 #define VENDOR_ATMEL_RFO	6
+#define VENDOR_WIFX		7
 
 #define HW_INFO_TOTAL_SIZE	0x20
 
