diff --git a/include/configs/at91-sama5_common.h b/include/configs/at91-sama5_common.h
index 30c6cd47ca..9c295d0dc7 100644
--- a/include/configs/at91-sama5_common.h
+++ b/include/configs/at91-sama5_common.h
@@ -46,33 +46,41 @@
 #define CONFIG_ENV_SIZE		0x4000
 #endif
 
+#ifndef CONFIG_BOOTCOMMAND
 #define CONFIG_BOOTCOMMAND	"if test ! -n ${dtb_name}; then "	\
 				    "setenv dtb_name at91-${board_name}.dtb; " \
 				"fi; "					\
 				"fatload mmc 0:1 0x21000000 ${dtb_name}; " \
 				"fatload mmc 0:1 0x22000000 zImage; "	\
 				"bootz 0x22000000 - 0x21000000"
+#endif
 
 #else
 
 #ifdef CONFIG_NAND_BOOT
 /* u-boot env in nand flash */
-#define CONFIG_ENV_OFFSET		0x140000
+#define CONFIG_ENV_OFFSET			0x140000
+#define CONFIG_ENV_SIZE				(128 * 1024) 			/* 128KiB */
 #define CONFIG_ENV_OFFSET_REDUND	0x100000
-#define CONFIG_ENV_SIZE			0x20000
-#define CONFIG_BOOTCOMMAND		"nand read 0x21000000 0x180000 0x80000;"	\
-					"nand read 0x22000000 0x200000 0x600000;"	\
-					"bootz 0x22000000 - 0x21000000"
+#define CONFIG_ENV_SIZE_REDUND		(CONFIG_ENV_SIZE)
+#define CONFIG_ENV_RANGE			(2 * CONFIG_ENV_SIZE)	/* Bad block handling */
+#ifndef CONFIG_BOOTCOMMAND
+#define CONFIG_BOOTCOMMAND			"nand read 0x21000000 0x180000 0x80000;"	\
+									"nand read 0x22000000 0x200000 0x600000;"	\
+									"bootz 0x22000000 - 0x21000000"
+#endif
 #elif CONFIG_SPI_BOOT
 /* u-boot env in serial flash, by default is bus 0 and cs 0 */
 #define CONFIG_ENV_OFFSET		0x6000
 #define CONFIG_ENV_SIZE			0x2000
 #define CONFIG_ENV_SECT_SIZE		0x1000
+#ifndef CONFIG_BOOTCOMMAND
 #define CONFIG_BOOTCOMMAND		"sf probe 0; "				\
 					"sf read 0x21000000 0x60000 0xc000; "	\
 					"sf read 0x22000000 0x6c000 0x394000; "	\
 					"bootz 0x22000000 - 0x21000000"
 #endif
+#endif
 
 #endif
 
